tools/check-links

```diff
@@ -0,0 +1,217 @@
+        self.visted_urls_soup = dict()
+        self.external_links_to_check = list()
+        self.broken_links = list()
+
+        try:
+            result = requests.get(self.site)
+            soup = BeautifulSoup(result.text, "html.parser")
+            self.urls = self.get_urls_from_page(soup)
+        except requests.exceptions.ConnectionError:
+            print("(error) failed to get", self.site)
+            sys.exit(1)
+
+
+
+class Crawler(initialize_crawler):
+    def __init__(self, site, allowed_file_ext_for_soup=allowed_file_ext_for_soup):
+        super().__init__(site, allowed_file_ext_for_soup)
+
+    def in_namespace(self, url) -> bool:
+        return urlparse(url).netloc == self.site_namespace
+
+    def get_urls_from_page(self, soup) -> list:
+        href_tags = soup.find_all(href=True)
+        hrefs = [tag.get("href") for tag in href_tags]
+        return hrefs
+
+    def in_allowed_file_soup(self, url) -> bool:
+        return urlparse(url).path.split(".")[-1] in self.allowed_file_ext_for_soup
+
+    def report_error(self, error, packet):
+        self.broken_links.append(packet.url)
+        print(f"({error}) {packet.parent}, {packet.url}")
+
+    def start_crawl(self):
+        self.crawl_and_report(self.site, self.urls)
+        self.check_external_links(self.external_links_to_check)
+
+    def get_urls_to_crawl(self, packet) -> list:
+        urls_from_page = list()
+        if self.in_allowed_file_soup(packet.url):
+            result = requests.get(packet.defraged_url)
+            self.visted_urls_soup[packet.defraged_url] = BeautifulSoup(
+                result.text, "html.parser"
+            )
+            if not check_fragment(
+                self.visted_urls_soup[packet.defraged_url], packet.fragment
+            ):
+                self.report_error("fragment missing", packet)
+            urls_from_page = self.get_urls_from_page(
+                self.visted_urls_soup[packet.defraged_url]
+            )
+        return urls_from_page
+
+    def check_internal_links(self, packet):
+        try:
+            _url = check_response(packet.defraged_url)
+            urls_from_page = self.get_urls_to_crawl(packet)
+            if len(urls_from_page) != 0:
+                self.crawl_and_report(_url, urls_from_page)
+        except UrlNotFound as e:
+            self.report_error(f"{e.response_code} error", packet)
+
+    def check_external_links(self, urls):
+        def is_url_broken(url_packet):
+            try:
+                check_response(url_packet.url)
+            except UrlNotFound as e:
+                self.report_error(f"{e.response_code} error", url_packet)
+
+        with Pool(max_workers=len(urls)) as executor:
+            future_to_url = {executor.submit(is_url_broken, packet) for packet in urls}
+            for future in future_to_url:
+                try:
+                    data = future.result()
+                except Exception as e:
+                    pass
```

Comment:
Almost-blocking: Catching `Exception` will hide a lot of bugs. Why are we catching exceptions here at all? When would we expect `is_url_broken` to raise an exception?